### ISSUE TITLE
Add option to the `prefer-template` rule to only warn on concat with variables

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#6572]
```
~/dev
❯ eslint --version
v2.13.1

~/dev
❯ echo "var key = 'fooBar';var foo = '--' + key.replace(/[A-Z]/g, '-$&').toLowerCase();" | eslint --stdin --rule='prefer-template:2'

<text>
  1:30  error  Unexpected string concatenation  prefer-template

✖ 1 problem (1 error, 0 warnings)
```
## Problem

The following:

``` js
var key = 'fooBar';
var foo = '--' + key.replace(/[A-Z]/g, '-$&').toLowerCase();
```

Would be very unreadable in a template string:

``` js
var key = 'fooBar';
var foo = `--${key.replace(/[A-Z]/g, '-$&').toLowerCase()}`;
```
## Request

I would like an option to only have template literals enforced when concatenating a string with variables, like:

``` js
var key = 'fooBar';
var foo = '--' + key;
```

⬇️ 

``` js
var key = 'fooBar';
var foo = `--${key}`;
```

**Comment by vitorbal** [comment]
So if anything other than variables is part of the concatenation, it would be considered valid?

```javascript
// valid
var foo = '--' + key.toLowerCase();

// invalid
var foo = '--' + key;
```

Think this makes sense, but just to play the devil's advocate for a bit, why wouldn't you just use an intermediate variable:

```javascript
var keyParsed = key.replace(/[A-Z]/g, '-$&').toLowerCase();
var foo = `--${keyParsed}`;
```

Doesn't that make it even more readable?

**Comment by platinumazure** [comment]
Between @vitorbal's suggestion of using an intermediate variable, the existence of inline config comments, and our new rule change guideline "Is the rule clearly incomplete without the requested enhancement?", I'm :-1: on this proposal.

**Comment by sindresorhus** [comment]
> So if anything other than variables is part of the concatenation, it would be considered valid?

@vitorbal Yes.

> Think this makes sense, but just to play the devil's advocate for a bit, why wouldn't you just use an intermediate variable:

The above is just a simple example. It's not always feasible or wanted to have intermediate variables.

The following:

```js
var foo = chalk.red('I') + ' like ' + chalk.blue('unicorns') + ' and ' + chalk.yellow('ponies');
```

Would have to be:

```js
var redI = chalk.red.bold('I');
var blueUnicorn = chalk.blue.bold('unicorns');
var yellowPonies = chalk.yellow.bold('ponies');
var foo = `${redI} like ${blueUnicorn} and ${yellowPonies}`;
```

Which is very unnecessarily verbose.

**Comment by platinumazure** [comment]
@sindresorhus I'm not really convinced by the new example-- seems to me that this is quite readable:

```js
var foo = `${chalk.red.bold('I')} like ${chalk.blue.bold('unicorns')} and ${chalk.yellow.bold('ponies')}`;
```

Seems to me it's honestly really case by case (because I agree the first example can be pretty hard to read). I just feel like assuming "variables" should be the only thing that should go in templates might be overly simplistic, certainly subjective (e.g., why not MemberExpressions, or maybe CallExpressions with no arguments?). I honestly still think at this point that your best bet is to use `// eslint-disable-line prefer-template` in those cases where you think a template is less readable.

**Comment by sindresorhus** [comment]
@platinumazure The reason I opened this request is because I would have to add too many `// eslint-disable-line prefer-template` for it to be feasible and I would just end up having to disable the rule for all my projects.

Another example of where ESLint wants me to use a template literal:

```js
const foo =  lines.map(x => {
	if (x.type === 'message') {
		return [
			'',
			x.severity === 'warning' ? logSymbols.warning : logSymbols.error,
			repeating(maxLineWidth - x.lineWidth) + chalk.dim(x.line + chalk.gray(':') + x.column),
			repeating(maxColumnWidth - x.columnWidth) + x.message,
			repeating(maxMessageWidth - x.messageWidth) + chalk.gray.dim(x.ruleId)
		].join('  ');
	}

	return '';
}).join('\n') + '\n\n';
```

It wants me to put all the above code (except the `const foo =  ` part) into a template literal, just because of the ` + '\n\n'`.

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
